### PR TITLE
Distribution Release API 

### PR DESCRIPTION
**Pull request by buptkang**
Since the backend file upload service has been updated, this PR updates the cli side to synchronize with the backend side.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -82,6 +82,8 @@
     "typescript": "^2.5.2"
   },
   "dependencies": {
+    "@types/progress": "^2.0.1",
```
Comment (christav) [inline package.json]: @types should be a devDependency, not a dependency.

File: src/commands/distribute/release.ts [hunk 2]
```diff
@@ -7,6 +7,12 @@ import * as Request from "request";
 import * as Path from "path";
 import * as Pfs from "../../util/misc/promisfied-fs";
 import { DefaultApp } from "../../util/profile";
+import FileUploadClient, { IFileUploadClientSettings, IProgress, IUploadResults, MessageLevel } from "appcenter-file-upload-client";
```
Comment (christav) [inline src/commands/distribute/release.ts]: What is this syntax with one name outside the braces and others inside? What does this do?

File: src/commands/distribute/release.ts [hunk 3]
```diff
@@ -39,50 +45,46 @@ export default class ReleaseBinaryCommand extends AppCommand {
   public releaseNotesFile: string;
 
   public async run(client: AppCenterClient): Promise<CommandResult> {
+
+    const bar = new ProgressBar(' Uploading a new release [:bar] :percent', {
```
Comment (christav) [inline src/commands/distribute/release.ts]: How does this progress bar integrate with the `--quiet` switch? Looks like it doesn't. You shouldn't display the progress bar if `--quiet` is given. Look at the stuff in the interaction directory (like out.text, out.progress, etc.) to see the patterns used to handle this.

File: package.json (hunk 1 above)
Comment (buptkang) [reply package.json]: Oh, good point.

File: src/commands/distribute/release.ts (hunk 2 above)
Comment (buptkang) [reply src/commands/distribute/release.ts]: FileUploadClient is the default export class, whereas the others are not. Here it is its definition: https://github.com/Microsoft/appcenter-file-upload-client/blob/master/src/index.ts

File: src/commands/distribute/release.ts [hunk 4]
```diff
@@ -144,83 +120,34 @@ export default class ReleaseBinaryCommand extends AppCommand {
         throw statusCode;
       }
     } catch (error) {
-      if (error === 404) {
-        throw failure(ErrorCodes.InvalidParameter, `distribution group ${this.distributionGroup} was not found`);
-      } else {
-        debug(`Failed to get users of distribution group ${this.distributionGroup}, returning null - ${inspect(error)}`);
-        return null;
-      }
+      throw failure(ErrorCodes.InvalidParameter, `distribution group ${this.distributionGroup} was not found`);
```
Comment (izikl) [inline src/commands/distribute/release.ts]: So if the server returns 500 we show a message "distribution group <name> was not found"? That will be incorrect and misleading. 
We should show a generic error message with a context ID (If context ID is shown when running with `--debug` in `client.distributionGroups.listUsers` then having a context ID printed here is just nice to have).

File: src/commands/distribute/release.ts [hunk 5]
```diff
@@ -144,83 +120,34 @@ export default class ReleaseBinaryCommand extends AppCommand {
...
     try {
-      abortReleaseUploadRequestResponse = await out.progress("Aborting release upload...",
-        clientRequest<models.ReleaseUploadEndResponse>((cb) => client.releaseUploads.complete(uploadId, app.ownerName, app.appName, "aborted", cb)));
+      uploadProgressResponse = await clientRequest<models.UploadProgress>((cb) => client.releaseUploads.get(assetId, 
+        app.ownerName, 
+        app.appName, 
+        cb));
     } catch (error) {
-      throw new Error(`HTTP ${abortReleaseUploadRequestResponse.response.statusCode} - ${abortReleaseUploadRequestResponse.response.statusMessage}`);
+      throw failure(ErrorCodes.Exception, `Processing the uploaded app file failed: ${error}.`);
```
Comment (izikl) [inline src/commands/distribute/release.ts]: I would change the error to "Failed to process the package." and again separate 500s from 400s. On 500s the the expectation from the user is ping us in intercom, on 400s we probably want them to change something in the package.

File: package.json [hunk 6]
```diff
@@ -82,6 +82,8 @@
...
+    "appcenter-file-upload-client": "^0.0.6",
```
Comment (sergey-akhalkov) [inline package.json]: Hi @buptkang, please change package version to `0.0.5`, I believe that `>= 0.0.6` is not going to work on Windows due to incorrect superagent dependency. `package.json` of  `appcenter-file-upload-client` should be updated to use the stable version of superagent.

Please let me know if you have any questions.

Comment (buptkang) [reply package.json]: Thanks for this information.

Comment (buptkang) [reply package.json]: Switched back to the previous version.

File: src/commands/distribute/release.ts (hunk 3 above)
Comment (buptkang) [reply src/commands/distribute/release.ts]: that is a really good point, have abstracted the method.

File: src/commands/distribute/release.ts [hunk 7]
```diff
@@ -39,50 +45,46 @@ export default class ReleaseBinaryCommand extends AppCommand {
...
+    // load release notes file or use provided release notes if none was specified
+    const releaseNotesString = await this.getReleaseNotesString();
 
-    debug("Creating release upload");
-    const createdReleaseUpload = await this.createReleaseUpload(client, app);
-    const uploadUri = createdReleaseUpload.uploadUrl;
-    const uploadId = createdReleaseUpload.uploadId;
+    // get number of distribution group users (and check distribution group existence)
+    // return null if request has failed because of any reason except non-existing group name.
+    const distributionGroupUsersNumber = await this.getDistributionGroupUsersNumber(client);
```
[the remark this answers is not in the document]
Comment (buptkang) [reply src/commands/distribute/release.ts]: Yeah, rollback to the original print result.

[the remark this answers is not in the document]
Comment (buptkang) [reply src/commands/distribute/release.ts]: Fixed.

**Comment by buptkang** [comment]
Close this PR, will create another one to map a special release version.

**Pull request closed without merging**